### PR TITLE
Fix #253: package confusion between conda and pip

### LINKED ISSUE #253: Potential package name confusion breaks conda-lock
Hi,

I spent a couple of hours today on creating an environment.yml files that pulls in `pytorch` packages from the `pytorch` channel:

```
dependencies:
  - conda-lock
  - cmake
  - pip
  - python=3.10
  - cudatoolkit==11.6
  - pytorch::ignite==0.4.9
  - pytorch::pytorch==1.12.1
  - pytorch::torchvision==0.13.1
  - pytorch::torchaudio==0.12.1
```

This seems to create a conda-lock.yml file without an issue. But if I add a pip dependencies section, and list a PIP package that depends on the `torch` package in PyPI, e.g.:

```
dependencies:
  - conda-lock
  - cmake
  - pip
  - python=3.10
  - cudatoolkit==11.6
  - pytorch::ignite==0.4.9
  - pytorch::pytorch==1.12.1
  - pytorch::torchvision==0.13.1
  - pytorch::torchaudio==0.12.1
  - pip:
    - pytorch_revgrad==0.2.0
```

I get the following exception:
```
  File "/data/ozan/mambaforge/lib/python3.10/site-packages/click/core.py", line 1130, in __call__                                                                                              
    return self.main(*args, **kwargs)                                                                                                                                                          
  File "/data/ozan/mambaforge/lib/python3.10/site-packages/click/core.py", line 1055, in main                                                                                                  
    rv = self.invoke(ctx)                                                                                                                                                                      
  File "/data/ozan/mambaforge/lib/python3.10/site-packages/click/core.py", line 1657, in invoke                                                                                                
    return _process_result(sub_ctx.command.invoke(sub_ctx))                                                                                                                                    
  File "/data/ozan/mambaforge/lib/python3.10/site-packages/click/core.py", line 1404, in invoke                                                                                                
    return ctx.invoke(self.callback, **ctx.params)                                                                                                                                             
  File "/data/ozan/mambaforge/lib/python3.10/site-packages/click/core.py", line 760, in invoke                                                                                                 
    return __callback(*args, **kwargs)                                                                                                                                                         
  File "/data/ozan/mambaforge/lib/python3.10/site-packages/click/decorators.py", line 26, in new_func                                                                                          
    return f(get_current_context(), *args, **kwargs)                                                                                                                                           
  File "/data/ozan/mambaforge/lib/python3.10/site-packages/conda_lock/conda_lock.py", line 1178, in lock                                                                                       
    lock_func(                                                                                                                                                                                 
  File "/data/ozan/mambaforge/lib/python3.10/site-packages/conda_lock/conda_lock.py", line 948, in run_lock                                                                                    
    make_lock_files(                                                                                                                                                                           
  File "/data/ozan/mambaforge/lib/python3.10/site-packages/conda_lock/conda_lock.py", line 388, in make_lock_files                                                                             
    lock_content = lock_content | create_lockfile_from_spec(                                                                                                                                   
  File "/data/ozan/mambaforge/lib/python3.10/site-packages/conda_lock/conda_lock.py", line 736, in create_lockfile_from_spec                                                                   
    deps = _solve_for_arch(                                                                                                                                                                    
  File "/data/ozan/mambaforge/lib/python3.10/site-packages/conda_lock/conda_lock.py", line 702, in _solve_for_arch
    pip_deps = solve_pypi(
  File "/data/ozan/mambaforge/lib/python3.10/site-packages/conda_lock/pypi_solver.py", line 310, in solve_pypi
    src_parser._apply_categories(requested=pip_specs, planned=planned)
  File "/data/ozan/mambaforge/lib/python3.10/site-packages/conda_lock/src_parser/__init__.py", line 297, in _apply_categories
    for dep in seperator_munge_get(planned, item).dependencies
  File "/data/ozan/mambaforge/lib/python3.10/site-packages/conda_lock/src_parser/__init__.py", line 288, in seperator_munge_get
    return d[key.replace("_", "-")]
KeyError: 'pytorch'
```

I believe this is due to the weird fact that:
 - In PyPI, pytorch package is called `torch` and `pytorch_revgrad` lists `torch` as a dependency.
 - However, in `pytorch` anaconda channel, the package is called `pytorch`

### PR DESCRIPTION
**Pull request by romain-intel**
Before this patch, the following would fail:
```
channels:
  - conda-forge
dependencies:
  - pip
  - dask
  - pip:
    - dask
```
The core issue seems to be that two Conda packages (`dask` and `dask-core`) both map to the `dask` pypi package. This was causing issues later on when assigning categories. This patch properly deals with multiple Conda packages per pip package and addresses the naming issue.

I have verified that the above package specification works as well as the others listed in #253.

### CODE REVIEW

**Comment by romain-intel** [comment]
I think there is some test flakiness given that it passes on other Python versions/platforms. Let me know if it is something else though.

**Comment by maresb** [comment]
pre-commit.ci autofix

**Comment by romain-intel** [comment]
Thanks for taking a look. I'll fix the pre-commit thing (I just need to extract a function out -- I had it in my first version and then undid it because I had forgotten why -- it was to remove the too complex function warning). I'll check the type issues too. May have missed something with the merge.

**Comment by maresb** [comment]
Awesome, thanks!!! I really want to try and get this in soon, but I still need to understand it first. Be warned there could be further merge conflicts.

**Comment by maresb** [comment]
Any thoughts on #322?

**Comment by maresb** [comment]
Closes #380

**Comment by romain-intel** [comment]
Sorry, I have been traveling -- I haven't forgotten. I'll try to get to it tonight or over the weekend.

**Comment by romain-intel** [comment]
I will see if I can add a test as well.

**Comment by romain-intel** [comment]
> Any thoughts on #322?

I must say that I am not fully well versed in the mapping rules. My gut feeling though is that @PanayotisManganaris is probably right in saying that there are probably cases where the mapping done will go wrong and it may be nice to offer users a "way out". There is currently no option to change to a different lookup table and/or to add to it. I could see it becoming quite possibly complicated though:
  - different lookup tables for different channels
  - ability for users to specify their own yaml lookup (that should be fairly straightforward to enable and may cover a large majority of these cases
  - ability for users to specify overrides in the yaml file to resolve (that may also cover a good variety of use cases).

**Comment by romain-intel** [comment]
OK, I added a simple test (nothing fancy) but hopefully this should be good to go. Apologies again for the delay -- it's been a very hectic few weeks but let me know if you need anything else. Would love to get this merged too (and released -- it's been a while since the last release :) ).

**Comment by maresb** [comment]
Thank you so much for this!!!

I really want to get this merged too, and I agree that we're due for a big release soon. 

I am also getting somewhat lost with how the mapping is supposed to work. Now that you have your head wrapped around this part of the logic, would you have any time to add a few comments? I find the code in this vicinity really difficult to read, not just this PR. Alternatively, maybe I could try an figure it out myself, and you could review what I write to see if it's accurate.

Some of the things I'm getting confused about: what is the `planned` mapping? In particular, what exactly are the keys, and why are the values `list`s of `LockedDependency`s? What is the general iteration strategy with `extract_planned_items`? And it what sense are they "extracted"? (Maybe we should be using a different word here?)

**Comment by romain-intel** [comment]
Let me see what I can add. I can't say I have a brilliant understanding of the code but can try to clarify some of the portions I understood. Hopefully with both of us we can come up with something sensible :).

**Comment by romain-intel** [comment]
hello -- apologies. I said I would add some comments and then the end of the quarter came. Now that it is done, your message came at the perfect time so I made some time to put at least some comments. I can't say that I fully understand why it is doing this category assignation and what the role of it is but hopefully it clarifies how it is doing it. Let me know if this helps and if you need something else.

**Comment by leeleavitt** [comment]
I composed 3 manual tests,

##### pyproject.toml

```
[build-system]
requires = ["poetry-core>=1.0.0"]
build-backend = "poetry.core.masonry.api"

[tool.poetry]
name = "tester"
version = "0.1.0"
description = "tester project"
requires-python = ">=3.9"

# Conda lock environment specs
[tool.conda-lock]
channels = [
    "bioconda",
    "conda-forge"
]
platforms = [
    "linux-64",
    "osx-64"
]

[tool.poetry.dependencies]
python = "3.9.12"
pip = "*"
pybedtools = "0.9.0"
miniwdl = "1.2.3"
pytest-wdl = {version = "*", source = "pypi"}
```

##### zenvironment-pytest-wdl.yaml
```
name: python
platforms:
  - linux-64
  - osx-64
channels:
  - bioconda
  - conda-forge
dependencies:
  - python=3.9.12
  - pip
  - miniwdl=1.2.3
  - pip:
    - pytest-wdl
```

##### zenvironment-python-tzdata.yaml
```
name: python
platforms:
  - linux-64
  - osx-64
channels:
  - bioconda
  - conda-forge
dependencies:
  - python=3.9.12
  - pybedtools
  - pip
  - pip:
    - pandas==1.3.5
 ```
 
 I first tested merging your branch into mine for #398, i did not get the errors. So then i removed my fix and tested my examples on your branch and i did not get the errors. 
 
 I went one step further and removed `_separator_munge_get` to simply,
 
 ```
 def _seperator_munge_get(
    d: Mapping[str, Union[List[LockedDependency], LockedDependency]], key: str
) -> Union[List[LockedDependency], LockedDependency]:
    # since separators are not consistent across managers (or even within) we need to do some double attempts here
    # try:
    return d[key]
```

And once again i did not get any error. Overall i believe this fixes my issues and i would love to see this get merged into main.

**Comment by maresb** [comment]
> I went one step further and removed `_separator_munge_get` to simply...

Awesome, this is really great info @leeleavitt! It would be great to be able to remove it since it's so confusing.

I also really appreciate the effort to design those tests. Did you implement them in PyTest? If so, would you be willing to put them in a PR? That would certainly help me with gathering the courage to merge this.

**Comment by leeleavitt** [comment]
> > I went one step further and removed `_separator_munge_get` to simply...
> 
> Awesome, this is really great info @leeleavitt! It would be great to be able to remove it since it's so confusing.
> 
> I also really appreciate the effort to design those tests. Did you implement them in PyTest? If so, would you be willing to put them in a PR? That would certainly help me with gathering the courage to merge this.

I was looking into the testing organization, but i could not see how to add these as tests. Do you have any pointers for where and how i could add these as tests? Since these would require network interaction i'm assuming I would need to create some cassettes?

**Comment by romain-intel** [comment]
Re `_separator_munge_get`: I did not change the function in my PR because I wasn't sure if there was a reason for it (chesterton's fence). I *believe* that it is there to make names that have _ or - the same as they are sometimes used interchangeably.

In terms of tests, I think you can add the .yml files just like I added the one test file for the one I ran. Is that what you are asking?

**Comment by lesteve** [comment]
> I believe that it is there to make names that have _ or - the same as they are sometimes used interchangeably.

In my experience pip does not care about `-` or `_` (or `.` by the way) e.g. `pip install scikit_learn` or `pip install scikit-learn`. I believe this kind of caveats should be supported in conda-lock as well so `_separator_munge_get` should stay ...

It would be really great to merge this, this would get rid of a lot of issues with mixed conda/pip dependencies.

**Comment by mariusvniekerk** [comment]
@maresb Added these tests in.  Will merge if they pass

**Comment by maresb** [comment]
Excellent, thanks a lot @mariusvniekerk for moving this forward!

**Comment by mariusvniekerk** [comment]
Probably worthwhile consolidating the various "solvability tests" at some point, but out of scope here

**Comment by maresb** [comment]
@mariusvniekerk I'd really like to get your thoughts on name canonicalization. For PyPI there is a [canonicalize_name](https://packaging.pypa.io/en/stable/utils.html#packaging.utils.canonicalize_name) function, so I think if we canonicalize PyPI names everywhere then there should no longer be a problem with those.

Then the only question is Conda, which to the best of my knowledge is trickier. I patched a bunch of the upstream Grayskull mapping stuff, and I'm wondering if there is any know example where `seperator_munge_get` actually does something?

**Comment by romain-intel** [comment]
Whoohoo! Thanks for moving this forward! If the tests don't pass, I'll take a look and see if there is something I can do.

File: conda_lock/pypi_solver.py [hunk 1]
```diff
@@ -228,6 +228,9 @@ def solve_pypi(
     for locked_dep in conda_locked.values():
         if locked_dep.name.startswith("__"):
             continue
+        # ignore packages that don't depend on Python
+        if locked_dep.manager != "pip" and "python" not in locked_dep.dependencies:
```
[the remark this answers is not in the document]
Comment (maresb) [reply conda_lock/pypi_solver.py:232]: That's intriguing!!! When does it trigger?

[the remark this answers is not in the document]
Comment (maresb) [reply conda_lock/pypi_solver.py:232]: What's an example of a non-Python pkg which causes trouble?

[the remark this answers is not in the document]
Comment (mariusvniekerk) [reply conda_lock/pypi_solver.py:232]: tzdata is the classic one

**Review by maresb — APPROVED** [review]

File: conda_lock/pypi_solver.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (maresb) [reply conda_lock/pypi_solver.py:232]: I'm concerned about the failing tests on Windows...

[the remark this answers is not in the document]
Comment (maresb) [reply conda_lock/pypi_solver.py:232]: The failures between the last two runs were disjoint. Maybe we just got unlucky with transient errors. Trying a third time.

**Comment by mariusvniekerk** [comment]
> Then the only question is Conda, which to the best of my knowledge is trickier. I patched a bunch of the upstream Grayskull mapping stuff, and I'm wondering if there is any know example where `seperator_munge_get` actually does something?  

Sadly conda packages have no canonical name form (practically many things are lowercase kebab, but it's not a hard rule and some things even use BOTH)

**Comment by romain-intel** [comment]
WHOOHOO!! Thanks for merging this!! Much appreciated! Now I can get off my custom build :)

**Comment by lesteve** [comment]
Great to see this merged, thanks to everyone involved, this will get rid of a lot of issues with mixed conda/pip dependencies!

**Pull request merged**